test/common/network/address_impl_test.cc

```diff
@@ -238,5 +238,58 @@ TEST(PipeInstanceTest, Basic) {
   EXPECT_EQ(nullptr, address.ip());
 }
 
+TEST(AddressFromSockAddr, IPv4) {
+  sockaddr_storage ss;
+  auto& sin = reinterpret_cast<sockaddr_in&>(ss);
+
+  sin.sin_family = AF_INET;
+  EXPECT_EQ(1, inet_pton(AF_INET, "1.2.3.4", &sin.sin_addr));
+  sin.sin_port = htons(6502);
+
+  EXPECT_THROW(addressFromSockAddr(ss, 1), EnvoyException);
+  EXPECT_THROW(addressFromSockAddr(ss, sizeof(sockaddr_in) - 1), EnvoyException);
+  EXPECT_THROW(addressFromSockAddr(ss, sizeof(sockaddr_in) + 1), EnvoyException);
+  auto addr1 = addressFromSockAddr(ss, sizeof(sockaddr_in));
+  EXPECT_EQ("1.2.3.4:6502", addr1->asString());
+
+  // Invalid family.
+  sin.sin_family = AF_UNSPEC;
+  EXPECT_THROW(addressFromSockAddr(ss, sizeof(sockaddr_in)), EnvoyException);
+}
+
+TEST(AddressFromSockAddr, IPv6) {
+  sockaddr_storage ss;
+  auto& sin6 = reinterpret_cast<sockaddr_in6&>(ss);
+
+  sin6.sin6_family = AF_INET6;
+  EXPECT_EQ(1, inet_pton(AF_INET6, "01:023::00Ef", &sin6.sin6_addr));
+  sin6.sin6_port = htons(32000);
+
+  EXPECT_THROW(addressFromSockAddr(ss, 1), EnvoyException);
+  EXPECT_THROW(addressFromSockAddr(ss, sizeof(sockaddr_in6) - 1), EnvoyException);
+  EXPECT_THROW(addressFromSockAddr(ss, sizeof(sockaddr_in6) + 1), EnvoyException);
+
+  auto addr1 = addressFromSockAddr(ss, sizeof(sockaddr_in6));
+  EXPECT_EQ("[1:23::ef]:32000", addr1->asString());
+}
+
+TEST(AddressFromSockAddr, Pipe) {
+  sockaddr_storage ss;
+  auto& sun = reinterpret_cast<sockaddr_un&>(ss);
+  sun.sun_family = AF_UNIX;
+  strcpy(sun.sun_path, "/some/path");
```

Comment:
Not that it matters much for tests, but I'm idly wondering if we can automatically forbid and flag dangerous standard library functions like `strcpy`...